machineconfiguration/v1/types_machineosconfig.go

```diff
@@ -0,0 +1,247 @@
+
+	// items contains a collection of MachineOSConfig resources.
+	Items []MachineOSConfig `json:"items"`
+}
+
+// MachineOSConfigSpec describes user-configurable options as well as information about a build process.
+type MachineOSConfigSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineConfigPoolReference `json:"machineConfigPool"`
+	// buildInputs is where user input options for the build live
+	// +kubebuilder:validation:Required
+	BuildInputs BuildInputs `json:"buildInputs"`
+	// buildOutputs holds all information needed to handle booting the image after a build
+	// This currently contains a currentImagePullSecret field, which should be provided if the final pull secret used to pull the image to nodes from the registry
+	// is different than the one used for pushing the image to the registry during the build.
+	// +optional
+	BuildOutputs *BuildOutputs `json:"buildOutputs,omitempty"`
+}
+
+// MachineOSConfigStatus describes the status this config object and relates it to the builds associated with this MachineOSConfig
+type MachineOSConfigStatus struct {
+	// observedGeneration represents the generation of the MachineOSConfig object observed by the Machine Config Operator's build controller.
+	// +kubebuilder:validation:XValidation:rule="self >= oldSelf || (self == 0 && oldSelf > 0)", message="observedGeneration must not move backwards except to zero"
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Required
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// currentImagePullspec is the fully qualified image pull spec used by the MCO to pull down the new OSImage. This must include sha256.
+	// The format of the image pullspec is:
+	// host[:port][/namespace]/name@sha256:<digest>
+	// The digest must be 64 characters long, and consist only of lowercase hexadecimal characters, a-f and 0-9.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=447
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@').size() == 2 && self.split('@')[1].matches('^sha256:[a-f0-9]{64}$'))`,message="the OCI Image reference must end with a valid '@sha256:<digest>' suffix, where '<digest>' is 64 characters long"
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@')[0].matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?/([a-zA-Z0-9-_]{0,61}/)?[a-zA-Z0-9-_.]*?$'))`,message="the OCI Image name should follow the host[:port][/namespace]/name format, resembling a valid URL without the scheme"
+	// +optional
+	CurrentImagePullspec string `json:"currentImagePullspec,omitempty"`
+	// machineOSBuild is a reference to the MachineOSBuild object for this MachineOSConfig, which contains the status for the image build
+	// +optional
+	MachineOSBuild *ObjectReference `json:"machineOSBuild,omitempty"`
+}
+
+// BuildInputs holds all of the information needed to trigger a build
+type BuildInputs struct {
```

Comment:
As a user of this API, how would I know what a BaseOSExtensions vs BaseOS vs Base image means?

I'm wondering if we can clean this API up a bit, there's a lot of, what appears to a naive eye at least, redundancy?
```
images:
  base: <>
  baseOS: <>
  baseOSExtensions: <>
pullSecret:
  name: <>
pushSecret:
  name: <>
builder:
  type: Job
  job:
    name: <>
pushSpec: <>
releaseVersion: <>
containerFiles: []
```
Or if we moved up to spec, some of this might look like
```
sourceImages:
  pullSecret: <> # For pulling these images
  base: <>
  baseOS: <>
  baseOSExtensions: <>
buildOutputs:
  pullSecret: 
    name: <> # For pulling the pushed image
  pushSecret:
    name: <> # For pushing the built image
  imagePushSpec: <>
builder:
  type: Job
  job:
    name: <>
releaseVersion: <>
containerFiles: []
```